Make file/file_list field preview images work with named sizes

Cai333 opened the pull request:
Fixes #844.

-  Add some utility methods for working with named images.

-  Filter the attachment meta data passed to the JS media uploader and adds all custom
  (i.e. registered with `add_image_size()`) image sizes.

-  File and File List fields try to get the closest named image size if passed an array for the `preview_size` parameter. If passed a string, the string is checked against existing image sizes. The named size is then passed as a data attribute to the hidden field to be used by the JS media uploader.

- The JavaScript that handles the media uploader now checks the named size passed from the hidden field and retrieves the correct image from the attachment object returned by the media uploader. The dimensions are then adjusted to fit within the preview size (rather than being stretched as they were before).

- Changed the default `preview_size` from 350x350px (for `file` fields) and 50x50px (for `file_list` fields) to the named sizes `large` and `thumbnail` respectively.

jtsternberg commented on includes/CMB2_Utils.php:
```diff
@@ -76,6 +76,99 @@ public static function image_id_from_url( $img_url ) {
+	 * @return false|string $data Named image size e.g. 'thumbnail'.
+	 */
+	public static function get_named_size( $size ) {
+		$image_sizes = self::get_available_image_sizes();
```
i'd move this inside the `if ( is_array( $size ) ) {` condition since it's not needed otherwise.

jtsternberg commented on includes/types/CMB2_Type_File.php:
```diff
@@ -28,7 +60,8 @@ public function render() {
 			'size'  => 45,
 			'desc'  => '',
-			'data-previewsize' => is_array( $img_size ) ? '[' . implode( ',', $img_size ) . ']' : 350,
+			'data-previewsize' => sprintf( '[%s,%s]', $size_width, $size_height ),
```
Couldn't this be `'[%d,%d]'`?

jtsternberg commented on includes/types/CMB2_Type_File_List.php:
```diff
@@ -55,7 +87,7 @@ public function render() {
 
 					$output .= $this->img_status_output( array(
-						'image'    => wp_get_attachment_image( $id, $img_size ),
+						'image'    => wp_get_attachment_image( $id, $img_size, null, array( 'class' => 'cmb-file_list-field-image', 'style' => 'height: auto;' ) ),
```
We should _probably_ abandon the inlines styles here.

Cai333 replied:
The 'auto' height is needed to stop the images being squashed if they're over the max-height that's set in the css, adding it here is probably just lazy though. I'll look at just adding it to the existing css.

jtsternberg commented on includes/CMB2_JS.php:
```diff
@@ -161,4 +161,48 @@ protected static function localize( $debug ) {
+	 * @param  array      $meta       Array of attachment meta data ( from wp_get_attachment_metadata() ).
+	 * @return string     filtered $response array
+	 */
+	public static function prepare_image_sizes_for_js( $response, $attachment, $meta ) {
```
It's tough to know the best place for these things, but I feel like this static method would be better suited for `CMB2_Type_File_Base`.

jtsternberg replied:
And to hook it in, I'd probably do something similar to the other field-specific stuff, like oembed ajax hooks:

```php
if ( in_array( $field['type'], array( 'file', 'file_list' ) ) ) {
	// Initiate attachment JS hooks
	add_filter( 'wp_prepare_attachment_for_js', 'CMB2_Type_File_Base::prepare_image_sizes_for_js', 10, 3 );
}
```

Cai333 replied:
Honestly, it was a struggle trying to decide where to hook this in, this sounds better!

jtsternberg commented on includes/types/CMB2_Type_File.php:
```diff
@@ -74,10 +107,9 @@ public function render() {
 
 				if ( $_id_value ) {
-					$image = wp_get_attachment_image( $_id_value, $img_size, null, array( 'class' => 'cmb-file-field-image' ) );
+					$image = wp_get_attachment_image( $_id_value, $img_size, null, array( 'class' => 'cmb-file-field-image', 'style' => 'height: auto;' ) );
```
We should _probably_ abandon the inlines styles here.

jtsternberg commented on includes/types/CMB2_Type_File.php:
```diff
@@ -74,10 +107,9 @@ public function render() {
 				} else {
-					$size = is_array( $img_size ) ? $img_size[0] : 350;
-					$image = '<img style="max-width: ' . absint( $size ) . 'px; width: 100%; height: auto;" src="' . $meta_value . '" alt="" />';
+					$image = '<img style="width: auto; height: auto;" src="' . $meta_value . '" class="cmb-file-field-image" alt="" />';
```
We should _probably_ abandon the inlines styles here.

jtsternberg commented on js/cmb2.js:
```diff
@@ -220,12 +221,41 @@ window.CMB2 = window.CMB2 || {};
+					
+					// Preview size dimensions
+					var previewWidth  = media.previewSize[0] ? media.previewSize[0] : 150;
+					var previewHeight = media.previewSize[1] ? media.previewSize[1] : 150;
```
Looks like you're changing this default as well (from 50 to 15), which again, is problematic for back-compatibility.

jtsternberg commented on js/cmb2.js:
```diff
@@ -220,12 +221,41 @@ window.CMB2 = window.CMB2 || {};
+					
+					// Get the correct dimensions and url if a named size is set and exists
+					// fallback to the 'large' size
+					if ( typeof( this.sizes[ media.sizeName ] ) !== 'undefined' ) {
```
use 'yoda' conditions

Cai333 commented:
Thanks for the review @jtsternberg 

Completely agree on all points; I've now moved all the repeating logic to utility methods. Your suggestion on hooking the filter makes much more sense too. Also reverted the changes to the defaults. Think I've addressed all your initial points, it's late though!

I moved the inline styles to the CSS file but  my compiled CSS is totally different, uses spaces instead of tabs etc (the diff basically shows every line as replaced) so I just commuted the sass file for now.

jtsternberg commented:
Awesome, thanks for the quick turnaround. I won't get a chance to test this till next week, but feel free to keep bumping this thread if I don't get around to it. I'd like to get this in the next release. No worries about the compiled assets. I can take care of those.

Cai333 commented:
@jtsternberg did you get a chance to test this at all?

I've been a bit snowed under myself but I'm going to try and do a bit more thorough testing on it in the next few days (I want to get this in a project I'm working on now).

jtsternberg commented:
Reviewing/testing this now. Please don't push any more changes as I'm starting to clean up, etc. If you have other changes to propose, let me know, and i'll push up the new branch to collaborate on.

jtsternberg commented:
Thanks a ton for your work @Cai333. This has been merged to trunk, and will be in the next release which I'll try to get out soon.

Cai333 commented:
No problem at all @jtsternberg
